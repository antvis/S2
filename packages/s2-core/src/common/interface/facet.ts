import type { Hierarchy } from '../../facet/layout/hierarchy';
import type { Node } from '../../facet/layout/node';
import type { ViewMeta } from './basic';

export interface AdjustLeafNodesParams {
  leafNodes: Node[];
  hierarchy: Hierarchy;
}

export interface ScrollChangeParams {
  offset: number;
  updateThumbOffset: boolean;
}

export type GetCellMeta = (
  rowIndex: number,
  colIndex: number,
) => ViewMeta | null;

export interface LayoutResult {
  /**
   * 列头节点, 对应 ColCell (含可视范围外)
   */
  colNodes: Node[];
  /**
   * 列头叶子节点, 对应 ColCell (含可视范围外)
   */
  colLeafNodes: Node[];
  /**
   * 列头节点层级结构 (含可视范围外)
   */
  colsHierarchy: Hierarchy;
  /**
   * 行头节点, 对应 RowCell (含可视范围外)
   */
  rowNodes: Node[];
  /**
   * 行头节点层级结构 (含可视范围外)
   */
  rowsHierarchy: Hierarchy;
  /**
   * 行头叶子节点, 对应 RowCell (含可视范围外)
   */
  rowLeafNodes: Node[];
  /**
   * 序号节点, 对应 SeriesNumberCell (含可视范围外)
   */
  seriesNumberNodes?: Node[];
  /**
   * 角头节点, 对应 CornerCell (含可视范围外)
   */
  cornerNodes?: Node[];
}
