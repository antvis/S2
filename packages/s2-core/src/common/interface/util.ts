export type DeepRequired<T extends Record<string, any>> = {
  [K in keyof T]-?: NonNullable<T[K]> extends Record<string, any>
    ? DeepRequired<NonNullable<T[K]>>
    : NonNullable<T[K]>;
};
