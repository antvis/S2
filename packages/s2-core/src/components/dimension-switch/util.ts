import { DimensionType, DimensionItem } from './dimension';

export interface OperatedType {
  [type: string]: string[];
}

export const getDimensionsByPredicate = (
  data: DimensionType[],
  predicate: (value: DimensionItem) => boolean,
) => {
  return data.reduce((obj, t) => {
    obj[t.type] = t.items.filter(predicate).map((i) => i.id);
    return obj;
  }, {} as OperatedType);
};
