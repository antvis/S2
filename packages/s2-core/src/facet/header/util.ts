import { isNil } from 'lodash';

export const getCellPadding = () => {
  const padding = [12, 4, 12, 4];
  const left = isNil(padding[3]) ? 0 : padding[3];
  const right = isNil(padding[1]) ? 0 : padding[1];
  const top = isNil(padding[0]) ? 0 : padding[0];
  const bottom = isNil(padding[2]) ? 0 : padding[2];
  return {
    left,
    right,
    top,
    bottom,
  };
};
