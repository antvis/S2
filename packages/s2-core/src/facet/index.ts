import { BaseFacet } from './base-facet';
import { PivotFacet } from './pivot-facet';
import { TableFacet } from './table-facet';

export { BaseFacet, PivotFacet, TableFacet };
