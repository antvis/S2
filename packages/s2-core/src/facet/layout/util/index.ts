export { default as getDimsConditionByNode } from './get-dims-condition-by-node';
export { default as getColHeight } from './get-col-height';
export { default as getColWidth } from './get-col-width';
export { default as getRowWidth } from './get-row-width';
export { default as processColsList } from './process-cols-list';
export { default as getHeaderHierarchyQuery } from './get-header-hierarchy-query';
