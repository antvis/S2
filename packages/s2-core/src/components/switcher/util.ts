import { filter, flatten, isEmpty, isNil, map } from 'lodash';
import { DraggableLocation } from 'react-beautiful-dnd';
import {
  FieldType,
  MAX_DIMENSION_COUNT,
  SWITCHER_PREFIX_CLS,
} from './constant';
import { SwitcherItem, SwitchResult, SwitchState } from './interface';
import { getClassNameWithPrefix } from '@/utils/get-classnames';

export const getSwitcherClassName = (...classNames: string[]) =>
  getClassNameWithPrefix(SWITCHER_PREFIX_CLS, ...classNames);

export const getNonEmptyFieldCount = (state: SwitchState) => {
  return [
    state[FieldType.Rows],
    state[FieldType.Cols],
    state[FieldType.Values],
  ].reduce((sum, value) => sum + (isEmpty(value) ? 0 : 1), 0);
};

export const getMainLayoutClassName = (nonEmptyCount: number) => {
  switch (nonEmptyCount) {
    case 1:
      return getSwitcherClassName('content', 'one-dimension');
    case 2:
      return getSwitcherClassName('content', 'two-dimensions');
    default:
      return getSwitcherClassName('content', 'three-dimensions');
  }
};

export const showDimensionCrossRows = (nonEmptyCount: number) =>
  nonEmptyCount < MAX_DIMENSION_COUNT;

export const isMeasureType = (fieldType: FieldType) =>
  fieldType === FieldType.Values;

export const moveItem = (
  source: SwitcherItem[],
  destination: SwitcherItem[],
  droppableSource: DraggableLocation,
  droppableDestination: DraggableLocation,
): SwitchState => {
  // change order in same column
  if (droppableDestination.droppableId === droppableSource.droppableId) {
    const updatingDestination = [...destination];
    const [removed] = updatingDestination.splice(droppableSource.index, 1);
    updatingDestination.splice(droppableDestination.index, 0, removed);
    return {
      [droppableDestination.droppableId]: updatingDestination,
    };
  }
  // move to other column
  const updatingSource = [...source];
  const updatingDestination = [...destination];

  const [removed] = updatingSource.splice(droppableSource.index, 1);
  updatingDestination.splice(droppableDestination.index, 0, removed);

  return {
    [droppableSource.droppableId]: updatingSource,
    [droppableDestination.droppableId]: updatingDestination,
  };
};

export const checkItem = (
  source: SwitcherItem[],
  checked: boolean,
  id: string,
  parentId?: string,
): SwitcherItem[] => {
  const target: SwitcherItem = {
    ...source.find((item) => item.id === (parentId ?? id)),
  };

  // 有 parentId 时，说明是第二层次项的改变
  if (parentId) {
    target.children = map(target.children, (item) => ({
      ...item,
      checked: item.id === id ? checked : item.checked,
    }));
  } else {
    target.checked = checked;
    target.children = map(target.children, (item) => ({
      ...item,
      checked: checked,
    }));
  }

  return source.map((item) => (item.id === target.id ? target : item));
};

export const generateSwitchResult = (state: SwitchState): SwitchResult => {
  // rows and cols can't be hidden
  const rows = map(state[FieldType.Rows], 'id');
  const cols = map(state[FieldType.Cols], 'id');

  // flatten all values and derived values
  const values = flatten(
    map(state[FieldType.Values], (item) => {
      const children = map(item.children, 'id');
      return [item.id, ...children];
    }),
  );

  const filterHiddenValues = (item: SwitcherItem) => item.checked === false;

  //  get all hidden values
  const hiddenValues = flatten(
    map(filter(state[FieldType.Values], filterHiddenValues), (item) => {
      const hiddenChildren = map(
        filter(item.children, filterHiddenValues),
        'id',
      );
      return [item.id, ...hiddenChildren];
    }),
  );

  return { rows, cols, values, hiddenValues };
};
