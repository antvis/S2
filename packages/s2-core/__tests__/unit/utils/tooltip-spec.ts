import { createFakeSpreadSheet } from 'tests/util/helpers';
import { BBox } from '@antv/g-canvas';
import { omit } from 'lodash';
import {
  getAutoAdjustPosition,
  setContainerStyle,
  getTooltipOptions,
} from '@/utils/tooltip';
import {
  CellTypes,
  getTooltipVisibleOperator,
  S2CellType,
  SpreadSheet,
  Tooltip,
  TOOLTIP_POSITION_OFFSET,
} from '@/index';
import { BaseFacet } from '@/facet/base-facet';

jest.mock('@/interaction/event-controller');

describe('Tooltip Utils Tests', () => {
  let s2: SpreadSheet;
  let tooltipContainer: HTMLDivElement;

  const containerSize = {
    width: 200,
    height: 200,
    top: 20,
    left: 10,
  };

  const tooltipSize = {
    width: 100,
    height: 100,
  };

  const bodySize = {
    width: 1000,
    height: 1000,
  };

  beforeAll(() => {
    s2 = createFakeSpreadSheet();
    tooltipContainer = {
      getBoundingClientRect: () =>
        ({
          ...tooltipSize,
        } as DOMRect),
    } as HTMLDivElement;
    s2.container.get = () => ({
      getBoundingClientRect: () => ({
        ...containerSize,
      }),
    });
    s2.facet = {
      panelBBox: {
        maxX: containerSize.width,
        maxY: containerSize.height,
      } as BBox,
    } as BaseFacet;
    document.body.getBoundingClientRect = () =>
      ({
        ...bodySize,
      } as DOMRect);
  });

  describe('Tooltip Position Tests', () => {
    test('should get original position if not set autoAdjustBoundary', () => {
      expect(
        getAutoAdjustPosition({
          position: { x: 0, y: 0 },
          autoAdjustBoundary: null,
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: TOOLTIP_POSITION_OFFSET.x,
        y: TOOLTIP_POSITION_OFFSET.y,
      });
    });

    test('should get original position if tooltip width less than container boundary', () => {
      expect(
        getAutoAdjustPosition({
          position: { x: 20, y: 20 },
          autoAdjustBoundary: 'container',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: 20 + TOOLTIP_POSITION_OFFSET.x,
        y: 20 + TOOLTIP_POSITION_OFFSET.y,
      });
    });

    test('should get auto adjust position if tooltip width more than container boundary', () => {
      // x
      expect(
        getAutoAdjustPosition({
          position: { x: 200, y: 20 },
          autoAdjustBoundary: 'container',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: containerSize.width - tooltipSize.width + containerSize.left,
        y: 20 + TOOLTIP_POSITION_OFFSET.y,
      });

      // y
      expect(
        getAutoAdjustPosition({
          position: { x: 20, y: 200 },
          autoAdjustBoundary: 'container',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: 20 + TOOLTIP_POSITION_OFFSET.x,
        y: containerSize.height - tooltipSize.height + containerSize.top,
      });

      // x, y
      expect(
        getAutoAdjustPosition({
          position: { x: 200, y: 200 },
          autoAdjustBoundary: 'container',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: containerSize.width - tooltipSize.width + containerSize.left,
        y: containerSize.height - tooltipSize.height + containerSize.top,
      });
    });

    test('should get auto adjust position if tooltip width more than body boundary', () => {
      // x
      expect(
        getAutoAdjustPosition({
          position: { x: 900, y: 20 },
          autoAdjustBoundary: 'body',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: bodySize.width - tooltipSize.width,
        y: 20 + TOOLTIP_POSITION_OFFSET.y,
      });

      // y
      expect(
        getAutoAdjustPosition({
          position: { x: 20, y: 900 },
          autoAdjustBoundary: 'body',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: 20 + TOOLTIP_POSITION_OFFSET.x,
        y: bodySize.height - tooltipSize.height,
      });

      // x, y
      expect(
        getAutoAdjustPosition({
          position: { x: 900, y: 900 },
          autoAdjustBoundary: 'body',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: bodySize.width - tooltipSize.width,
        y: bodySize.height - tooltipSize.height,
      });
    });

    test('should get auto adjust position if tooltip width more than container boundary and real render width less container width', () => {
      const panelBBox = {
        maxX: 120,
        maxY: 120,
      };

      s2.facet = {
        panelBBox: panelBBox,
      } as BaseFacet;

      // x, y
      expect(
        getAutoAdjustPosition({
          position: { x: 100, y: 100 },
          autoAdjustBoundary: 'container',
          spreadsheet: s2,
          tooltipContainer,
        }),
      ).toEqual({
        x: panelBBox.maxX - tooltipSize.width + containerSize.left,
        y: panelBBox.maxY - tooltipSize.height + containerSize.top,
      });
    });
  });

  describe('Get Tooltip Options Tests', () => {
    const getCellNameByType = (cellType: CellTypes) => {
      return {
        [CellTypes.ROW_CELL]: 'row',
        [CellTypes.COL_CELL]: 'col',
        [CellTypes.DATA_CELL]: 'data',
        [CellTypes.CORNER_CELL]: 'corner',
      }[cellType];
    };

    test.each([
      CellTypes.ROW_CELL,
      CellTypes.COL_CELL,
      CellTypes.DATA_CELL,
      CellTypes.CORNER_CELL,
    ])(
      'should use %o tooltip content from tooltip config first for string content',
      (cellType) => {
        const tooltipContent = `${cellType} tooltip content`;
        const defaultTooltipContent = 'default tooltip content';
        const type = getCellNameByType(cellType);

        const tooltip: Tooltip = {
          content: defaultTooltipContent,
          [type]: {
            content: tooltipContent,
          },
        };

        const spreadsheet = {
          getCellType: () => cellType,
          options: {
            tooltip,
          },
        } as unknown as SpreadSheet;

        expect(getTooltipOptions(spreadsheet, {} as Event)).toEqual({
          content: tooltipContent,
          [type]: {
            content: tooltipContent,
          },
        });
      },
    );

    test.each([
      CellTypes.ROW_CELL,
      CellTypes.COL_CELL,
      CellTypes.DATA_CELL,
      CellTypes.CORNER_CELL,
    ])(
      'should use %o tooltip options and merge base tooltip config',
      (cellType) => {
        const type = getCellNameByType(cellType);

        const tooltip: Tooltip = {
          showTooltip: false,
          content: '',
          operation: {
            hiddenColumns: true,
            trend: true,
            sort: true,
            tableSort: true,
            menus: [{ key: 'menu-a', text: 'menu-a' }],
          },
          [type]: {
            showTooltip: true,
            operation: {
              hiddenColumns: false,
              menus: [{ key: 'menu-b', text: 'menu-b' }],
            },
          },
        };

        const spreadsheet = {
          getCellType: () => cellType,
          options: {
            tooltip,
          },
        } as unknown as SpreadSheet;

        const tooltipOptions = omit(
          getTooltipOptions(spreadsheet, {} as Event),
          [type],
        );
        expect(tooltipOptions).toEqual({
          showTooltip: true,
          content: '',
          operation: {
            hiddenColumns: false,
            trend: true,
            sort: true,
            tableSort: true,
            menus: [{ key: 'menu-b', text: 'menu-b' }],
          },
        });
      },
    );

    test('should filter not displayed tooltip operation menus', () => {
      const mockCell = {
        cellType: CellTypes.DATA_CELL,
      } as unknown as S2CellType;
      const onClick = jest.fn();

      const defaultMenus = [
        {
          key: 'default-menu',
          text: 'default-menu',
        },
      ];

      const operation: Tooltip['operation'] = {
        onClick,
        menus: [
          { key: 'menu-0', text: '默认显示(未声明visible属性)' },
          { key: 'menu-1', text: '默认显示', visible: true },
          { key: 'menu-2', text: '默认隐藏', visible: false },
          { key: 'menu-3', text: '动态始终显示', visible: () => true },
          { key: 'menu-4', text: '动态始终显示', visible: () => false },
          {
            key: 'menu-5',
            text: '动态显示',
            visible: (cell) => cell.cellType === CellTypes.DATA_CELL,
          },
          {
            key: 'menu-6',
            text: '动态隐藏',
            visible: (cell) => cell.cellType !== CellTypes.DATA_CELL,
          },
          {
            key: 'menu-7',
            text: '父节点显示, 子节点隐藏',
            visible: true,
            children: [
              {
                key: 'menu-7-1',
                text: '父节点显示, 子节点隐藏',
                visible: false,
              },
            ],
          },
          {
            key: 'menu-8',
            text: '父节点隐藏, 子节点显示 (应该父,子都不显示)',
            visible: false,
            children: [
              {
                key: 'menu-8-1',
                text: '父节点隐藏, 子节点显示',
                visible: true,
              },
            ],
          },
        ],
      };
      const operator = getTooltipVisibleOperator(operation, {
        cell: mockCell,
        defaultMenus,
      });
      const visibleSubMenus = operator.menus.find(
        ({ key }) => key === 'menu-7',
      );

      expect(operator.onClick).toEqual(onClick);
      expect(operator.menus.map(({ key }) => key)).toEqual([
        'default-menu',
        'menu-0',
        'menu-1',
        'menu-3',
        'menu-5',
        'menu-7',
      ]);
      expect(visibleSubMenus.children).toHaveLength(0);
    });
  });

  test('should set container style', () => {
    const container = document.createElement('div');

    setContainerStyle(container, {
      style: {
        width: '100px',
        pointerEvents: 'none',
      },
    });

    expect(container.style.width).toEqual('100px');
    expect(container.style.pointerEvents).toEqual('none');
  });

  test('should set container class name', () => {
    const container = document.createElement('div');

    setContainerStyle(container, {
      className: 'test',
    });

    expect(container.className).toEqual('test');
  });
});
