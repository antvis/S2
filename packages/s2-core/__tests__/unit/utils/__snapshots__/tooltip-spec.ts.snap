// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip Utils Tests Tooltip Get Data Tests Tooltip Description Tests should get col cell descriptions 1`] = `
Object {
  "description": "类别说明。。",
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests Tooltip Description Tests should get data cell description 1`] = `
Object {
  "description": "数量说明。。",
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests Tooltip Description Tests should get row cell description 1`] = `
Object {
  "description": "省份说明。。",
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [
    Object {
      "name": "数量",
      "originValue": 15420,
      "selectedData": Array [
        CellData {
          "extraField": "number",
          "raw": Object {
            "city": "杭州市",
            "number": 7789,
            "province": "浙江省",
            "sub_type": "桌子",
            "type": "家具",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "city": "杭州市",
            "number": 5343,
            "province": "浙江省",
            "sub_type": "沙发",
            "type": "家具",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "city": "杭州市",
            "number": 945,
            "province": "浙江省",
            "sub_type": "笔",
            "type": "办公用品",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "city": "杭州市",
            "number": 1343,
            "province": "浙江省",
            "sub_type": "纸张",
            "type": "办公用品",
          },
        },
      ],
      "value": 15420,
    },
  ],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests should get correctly data cell details 1`] = `
Object {
  "description": undefined,
  "details": Array [
    Object {
      "name": "number",
      "value": 7789,
    },
  ],
  "headInfo": Object {
    "cols": Array [],
    "rows": Array [
      Object {
        "name": undefined,
        "value": "-",
      },
    ],
  },
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests should get custom tree row cell summary data 1`] = `
Object {
  "description": "指标1描述",
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [
    Object {
      "name": "指标1",
      "selectedData": Array [
        CellData {
          "extraField": "measure-1",
          "raw": Object {
            "measure-1": 13,
            "measure-2": 2,
            "measure-3": 3,
            "measure-4": 4,
            "measure-5": 5,
            "measure-6": 6,
            "sub_type": "桌子",
            "type": "家具",
          },
        },
        CellData {
          "extraField": "measure-1",
          "raw": Object {
            "measure-1": 11,
            "measure-2": 8,
            "measure-3": 32,
            "measure-4": 43,
            "measure-5": 45,
            "measure-6": 65,
            "sub_type": "椅子",
            "type": "家具",
          },
        },
      ],
      "value": 24,
    },
  ],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests should get custom tree row root cell summary data 1`] = `
Object {
  "description": "a-1 描述",
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [
    Object {
      "name": "自定义节点 a-1",
      "selectedData": Array [
        Object {
          "$$extra$$": "a-1",
          "sub_type": "桌子",
          "type": "家具",
        },
        Object {
          "$$extra$$": "a-1",
          "sub_type": "椅子",
          "type": "家具",
        },
      ],
      "value": null,
    },
  ],
  "tips": undefined,
}
`;

exports[`Tooltip Utils Tests Tooltip Get Data Tests should get grand totals row cell summary data 1`] = `
Object {
  "description": undefined,
  "details": null,
  "headInfo": null,
  "infos": undefined,
  "interpretation": undefined,
  "name": null,
  "summaries": Array [
    Object {
      "name": "数量",
      "originValue": 78868,
      "selectedData": Array [
        CellData {
          "extraField": "number",
          "raw": Object {
            "number": 26193,
            "sub_type": "桌子",
            "type": "家具",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "number": 23516,
            "sub_type": "沙发",
            "type": "家具",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "number": 12321,
            "sub_type": "笔",
            "type": "办公用品",
          },
        },
        CellData {
          "extraField": "number",
          "raw": Object {
            "number": 16838,
            "sub_type": "纸张",
            "type": "办公用品",
          },
        },
      ],
      "value": 78868,
    },
  ],
  "tips": undefined,
}
`;
